File: rnn/scala/rnn.scala
class RNN(inputDim: Int, hiddenDim: Int, outputDim: Int) {

  val weights = Array.fill(inputDim)(Array.fill(hiddenDim)(scala.util.Random.nextDouble - 0.5))
  val biases = Array.fill(hiddenDim)(scala.util.Random.nextDouble - 0.5)

  val h = Array.fill(hiddenDim)(0.0)

  def train(data: Array[Array[Double]]): Unit = {
    for (sequence <- data) {
      for (t <- sequence) {
        val z = 0.0
        for (i <- 0 until inputDim) {
          z += t(i) * weights(i)(h.size)
        }
        h = h :+ scala.math.tanh(z + biases(h.size))
      }
    }
  }

  def predict(data: Array[Double]): Array[Double] = {
    val predictions = Array.ofDim[Double](data.length)
    h = Array.fill(hiddenDim)(0.0)
    for (t <- data) {
      val z = 0.0
      for (i <- 0 until inputDim) {
        z += t(i) * weights(i)(h.size)
      }
      h = h :+ scala.math.tanh(z + biases(h.size))
      predictions(t.length - 1) = h(h.length - 1)
    }
    return predictions
  }
}

File: rnn/javascript/rnn.js
function RNN(inputDim, hiddenDim, outputDim) {
    this.inputDim = inputDim;
    this.hiddenDim = hiddenDim;
    this.outputDim = outputDim;

    this.weights = [];
    for (let i = 0; i < inputDim; i++) {
        this.weights.push([]);
        for (let j = 0; j < hiddenDim; j++) {
            this.weights[i].push(Math.random() - 0.5);
        }
    }

    this.biases = [];
    for (let i = 0; i < hiddenDim; i++) {
        this.biases.push(Math.random() - 0.5);
    }

    this.h = [];
}

RNN.prototype.train = function(data) {
    for (let sequence of data) {
        this.h = [];
        for (let t of sequence) {
            let z = 0;
            for (let i = 0; i < inputDim; i++) {
                z += t * this.weights[i][this.h.length];
            }
            this.h.push(Math.tanh(z + this.biases[this.h.length]));
        }
    }
};

RNN.prototype.predict = function(data) {
    let predictions = [];
    this.h = [];
    for (let t of data) {
        let z = 0;
        for (let i = 0; i < inputDim; i++) {
            z += t * this.weights[i][this.h.length];
        }
        this.h.push(Math.tanh(z + this.biases[this.h.length]));
        predictions.push(this.h[this.h.length - 1]);
    }
    return predictions;
};
